Tweaking error overlay styles

bvaughn opened the pull request:
Copied from #2191

This is kind of quick and dirty. Basically just editing style values inline.

I'd be happy to brainstorm ways to clean up some of the styling to make it more maintainable if you're interested? 😁

### Demos

##### Scrolling demo

<img alt="Scrolling demo" src="https://cloud.githubusercontent.com/assets/29597/26160088/986388a0-3b20-11e7-9cc9-1a0489cf4c3e.gif" width="318" height="567">

##### Chrome

<img width="432" height="702" alt="Chrome screenshot" src="https://cloud.githubusercontent.com/assets/29597/26160213/ebda3574-3b20-11e7-8f35-b58a7777198f.png">

##### iPhone

<img width="485" height="303" alt="Phone screenshot" src="https://cloud.githubusercontent.com/assets/29597/26160170/d0cc2080-3b20-11e7-9317-f8d0ed50b166.png">

gaearon commented:
I hesitated about going completely full screen but I think I’m on board because:

* Often the page underneath is blank, and it feels jarring to have an overlay on top of it appear. The fullscreen version feels more natural in this case.
* With Fiber, the whole tree gets unmounted on error (unless you define boundaries) so it's going to be blank in even more cases. Even if today you usually see the app underneath, in React 16 not so much.

I really like that the error message stays on screen at all times. This makes traversing the stack feel less like an aimless journey and more like figuring out what happened.

gaearon commented:
For the reference, the update looks like this:
![screen shot 2017-05-17 at 22 16 34](https://cloud.githubusercontent.com/assets/810438/26176592/9450c024-3b4e-11e7-881c-a026456099ea.png)
![screen shot 2017-05-17 at 22 16 38](https://cloud.githubusercontent.com/assets/810438/26176593/9458c198-3b4e-11e7-92c5-54dbfe9ac0e0.png)

Timer commented:
Maybe:
> It will not appear **if** the application crashes in production.

😄

gaearon commented:
Here’s specific, hopefully constructive feedback on things that I would like to see fixed to proceed with this.

## Issues

### Missing Multiple Error Navigation

We need to fix this case:

```js
[1, 2, 3].forEach((i) => setTimeout(() => {
  throw new Error(i)
}));
```

The errors are still navigatable with arrows, but there used to be a bar saying "Error X of Y" and navigation buttons (that hinted at arrows as shortcuts).

### Unintentional Color

From our conversation I gathered you were surprised that there are two selection colors:
<img width="522" alt="screen shot 2017-05-17 at 22 45 03" src="https://cloud.githubusercontent.com/assets/810438/26177665/7f49da22-3b52-11e7-9b27-98aa2318a863.png">

This is intentional because I want to differentiate the line that has thrown from merely frames leading up to it up in the call stack. It helps beginners realize which line to look at most closely.

However I realize you might want to tweak the exact color we use for non-error frames. In fact perhaps it would be best to make the background of non-error frames more neutral too so that they don’t look as nasty as the ones that have actually thrown.

### Selected Line Alignment

Not sure if it was an issue before or not, but I noticed the selection looks a tiny bit below the text:
<img width="1005" alt="screen shot 2017-05-17 at 22 47 21" src="https://cloud.githubusercontent.com/assets/810438/26177733/d939245c-3b52-11e7-85a1-84f4419ac34d.png">

Not sure if this is even fixable, or just how text works in browsers.
I would expect the highlighting to match up with the `|` character but maybe I have wrong expectations.

### Syntax Overlay Parity

It is jarring to switch between this:
![screen shot 2017-05-17 at 22 55 52](https://cloud.githubusercontent.com/assets/810438/26178002/07e41586-3b54-11e7-9dca-1c660ab30a4c.png)


and this:

![screen shot 2017-05-17 at 22 56 01](https://cloud.githubusercontent.com/assets/810438/26178006/0d2ac3d2-3b54-11e7-8c39-a9cfebe538b5.png)

We can start by using the same font sizes and positions for headers.

Ideally, I’m thinking that we should parse the code frame out of the message when it exists with a regex, and wrap it into a separate `<pre>` with styles identical to the runtime overlay. However I recognize this is more effort, and we can postpone that. 

## Nits

These are my personal preferences.

### Header Bottom Margin

Compared to other vertical margins on the page, header bottom margin seems too small.
For example it is smaller than the margin between "stack frame expanded" and next frame:

<img width="494" alt="screen shot 2017-05-17 at 22 33 25" src="https://cloud.githubusercontent.com/assets/810438/26177426/73296d58-3b51-11e7-8ef6-7ced8edf41fd.png">

Can we make it a bit larger? Maybe like this:

<img width="530" alt="screen shot 2017-05-17 at 22 38 38" src="https://cloud.githubusercontent.com/assets/810438/26177465/97d3e1f6-3b51-11e7-9b4a-6a951811e8a8.png">


### Scrollable Header

On narrow widths, header becomes scrollable which caught be my surprised. I missed a part of the message.

![](https://d2ppvlu71ri8gs.cloudfront.net/items/2d032z2y2n1y381j0w1s/Screen%20Recording%202017-05-17%20at%2022.27.gif?v=2a2e26c8)

### Top Header Margin

The header text looks a bit too close to the window top for my taste.

<img width="243" alt="screen shot 2017-05-17 at 22 22 05" src="https://cloud.githubusercontent.com/assets/810438/26176808/5e08ec02-3b4f-11e7-8988-e15ab21b4337.png">

This is subjective but I would feel more comfortable if it was moved down just a tiny bit more.
At least when we're large enough to see the automargins.
Maybe like this:

<img width="1116" alt="screen shot 2017-05-17 at 22 26 19" src="https://cloud.githubusercontent.com/assets/810438/26176980/e4babcee-3b4f-11e7-8066-0b168396fc18.png">

gaearon commented:
I pushed a temporary fix that reintroduces "N errors on the page". It flickers when switching from one to many errors now, but I won't adjust it as I'm not sure how you'd like to change the design for this case.
<img width="288" alt="screen shot 2017-05-17 at 23 54 02" src="https://cloud.githubusercontent.com/assets/810438/26179515/1ffe0732-3b5c-11e7-8750-66bcd251c1c5.png">

gaearon commented:
I think this is good enough for now, and I’ll get it in. Thank you!
IMO some of the things above are still worth addressing but I want to cut another canary so merging.

bvaughn commented:
> On narrow widths, header becomes scrollable which caught me by surprise. I missed a part of the message.

FWIW this is intentional, so really long error messages don't totally consume all of the space on a small screen (like mobile). That wouldn't be a good use case b'c you wouldn't be able to see the stack at all. So I set a max height to 35%, with scroll.

bvaughn commented:
> Compared to other vertical margins on the page, header bottom margin seems too small.
For example it is smaller than the margin between "stack frame expanded" and next frame:

Vertical space is very precious on smaller screens, and the choice to use inline styles for this project makes @media queries a pain in the butt to use.

Honestly, maybe we should refactor these styles to just be CSS rather than inline style objects.

bvaughn commented:
> From our conversation I gathered you were surprised that there are two selection colors

Yes and no. I was mostly just half-asleep. I specifically tested the error case that showed the 2nd highlighting style but the yellow looked a little different to me for some reason in your screenshot. Chalk it up to sleepiness.

The suggestion of possibly changing the background color seems good though. I'll play with that.

bvaughn commented:
> The errors are still navigatable with arrows, but there used to be a bar saying "Error X of Y" and navigation buttons (that hinted at arrows as shortcuts).
> **NOTE: I pushed a partial fix that makes it visible but we probably want to tweak the design.**

Hm. After fetching+pulling, the styles are broken for the nav. (It's way above the fold, not visible, due to the negative 4m top margin.) Did you forget to commit/push something?

gaearon commented:
>I wonder if we should put the stepper above the title?

Up to you! I just wasn't sure where it would make sense.

bvaughn commented:
How do you feel about this alternate style for secondary errors?

<img width="1033" alt="screen shot 2017-05-18 at 11 40 02 am" src="https://cloud.githubusercontent.com/assets/29597/26196272/c8bddee4-3bbe-11e7-8f3f-e169ac2dcba3.png">

bvaughn commented:
> ### Selected Line Alignment
> Not sure if it was an issue before or not, but I noticed the selection looks a tiny bit below the text:

I haven't seen that misalignment. Which browser? For me it aligns correctly:

#### Chrome
<img width="424" alt="screen shot 2017-05-18 at 11 42 01 am" src="https://cloud.githubusercontent.com/assets/29597/26196363/090743d2-3bbf-11e7-946a-ddd180cf1058.png">

#### Firefox
<img width="427" alt="screen shot 2017-05-18 at 11 43 12 am" src="https://cloud.githubusercontent.com/assets/29597/26196428/30f82e1a-3bbf-11e7-8a9b-46a7bcfe8cdc.png">

#### Safari
<img width="416" alt="screen shot 2017-05-18 at 11 43 02 am" src="https://cloud.githubusercontent.com/assets/29597/26196431/346885a4-3bbf-11e7-870a-cde8d5bf9a5c.png">

#### iOS Safari
<img width="314" alt="screen shot 2017-05-18 at 11 43 34 am" src="https://cloud.githubusercontent.com/assets/29597/26196448/3eb7adaa-3bbf-11e7-8b6c-b19fdab3e53a.png">

bvaughn commented:
> ### Browser Compatibility
> It is important that the overlay displays well on IE9+ because it’s annoying to debug errors if the tool supposed to help you breaks by itself. I haven’t tested it yet but I can tomorrow.

Unfortunately I don't have an IE 9 or 10 VM installed anymore b'c I ran out of disc space a while ago. I'd download one, but it would take days on the conference VM. Maybe you can check this initially and let me know if there are problems?

bvaughn commented:
I think I addressed the non-critical things and some of the nits, including:

- [x] Moved margin between header and file name to header, so when content was scrolled, the header would remain more separate
- [x] Made build-time and runtime overlays better match
- [x] Secondary error `<pre>` style now uses yellow bg instead of red
- [x] "Scrollable Header" (see above comment to why this is necessary) but I did increase the `max-height` from 35% to 50%.
- [x] Fixed header and "X" button vertical alignment

Did not address:

- [ ] Selected Line Alignment (works for me)
- [ ] IE testing (no VM)

##### Runtime error (iOS)

<img width="321" alt="screen shot 2017-05-18 at 12 05 12 pm" src="https://cloud.githubusercontent.com/assets/29597/26197463/89cf3878-3bc2-11e7-9385-c065e17b08ba.png">

##### Runtime error (desktop)

<img width="1054" alt="screen shot 2017-05-18 at 12 05 27 pm" src="https://cloud.githubusercontent.com/assets/29597/26197412/637cc726-3bc2-11e7-9ef7-746faf9c2129.png">

##### Build error (iOS)

<img width="320" alt="screen shot 2017-05-18 at 12 09 33 pm" src="https://cloud.githubusercontent.com/assets/29597/26197547/e884fb3c-3bc2-11e7-942d-a10cc2a0a723.png">

##### Build error (desktop)

<img width="1041" alt="screen shot 2017-05-18 at 12 09 26 pm" src="https://cloud.githubusercontent.com/assets/29597/26197554/f2533dfe-3bc2-11e7-895e-608becec3a4e.png">

gaearon commented:
>How do you feel about this alternate style for secondary errors?

Looks good!

>I haven't seen that misalignment. 

If it looks good to you then it's good. I was confused by `|` right after the line number not having the whole line's height but maybe it's just not how it's supposed to work.

>Maybe you can check this initially and let me know if there are problems?

Will do later today!

gaearon commented:
Can you please send a follow up PR to test changes?

bvaughn commented:
yeah, sry. Was pushing. (internet here is slow)
